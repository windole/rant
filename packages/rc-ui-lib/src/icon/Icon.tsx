import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
