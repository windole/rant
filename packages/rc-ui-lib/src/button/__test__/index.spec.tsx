import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../../icon';
import Button from '..';

describe('Button', () => {
  it('renders correctly', () => {
    expect(<Button />).toMatchRenderedSnapshot();
  });

  it('mount correctly', () => {
    expect(() => mount(<Button />)).not.toThrow();
  });

  it('should render empty button without errors', () => {
    const wrapper = mount(
      <Button>
        {null}
        {undefined}
      </Button>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should simulate click event', () => {
    const onClick = jest.fn();
    const wrapper = mount(<Button onClick={onClick} />);
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should not emit click event when disabled', () => {
    const onClick = jest.fn();
    const wrapper = mount(<Button disabled onClick={onClick} />);
    wrapper.simulate('click');
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should render correctly when loading prop is set', () => {
    const wrapper = mount(<Button type="primary" loading />);
    expect(wrapper.find('.rc-loading__spinner').exists()).toBeTruthy();
  });

  it('should hide border when color is gradient', () => {
    const wrapper = mount(<Button color="linear-gradient(#000, #fff)" />);
    expect(wrapper.getDOMNode().style.border).toEqual('0px');
  });

  it('should hide border when color is not gradient', () => {
    const wrapper = mount(<Button type="primary" color="red" />);
    expect(wrapper.getDOMNode().style.borderColor).toEqual('red');
  });

  it('should change icon class prefix when using icon-prefix prop', () => {
    const wrapper = mount(<Button icon="success" iconPrefix="my-icon" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly when icon is JSX element', () => {
    const wrapper = mount(<Button icon={<Icon name="shop-o" />} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render iconPosition correctly', () => {
    const wrapper = mount(<Button type="primary" iconPosition="right" />);
    expect(wrapper.find('.rc-loading__spinner').exists()).toBeFalsy();
  });

  it('should render buttton group correctly', () => {
    const wrapper = mount(
      <Button.Group>
        <Button type="primary">next step</Button>
      </Button.Group>,
    );
    expect(wrapper.find('.rc-button-group').exists()).toBeTruthy();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
