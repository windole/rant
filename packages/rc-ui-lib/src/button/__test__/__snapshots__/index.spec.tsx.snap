// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  class="rc-button rc-button--default rc-button--normal"
>
  <div
    class="rc-button__content"
  />
</button>
`;

exports[`Button should change icon class prefix when using icon-prefix prop 1`] = `
<Button
  icon="success"
  iconPosition="left"
  iconPrefix="my-icon"
  size="normal"
  tag="button"
  type="default"
>
  <button
    className="rc-button rc-button--default rc-button--normal"
    onClick={[Function]}
    style={Object {}}
  >
    <div
      className="rc-button__content"
    >
      <Icon
        className="rc-button__icon"
        classPrefix="my-icon"
        name="success"
        tag="i"
      >
        <Badge
          className="rc-button__icon my-icon my-icon-success"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-button__icon my-icon my-icon-success"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
  </button>
</Button>
`;

exports[`Button should render buttton group correctly 1`] = `
<ButtonGroup>
  <div
    className="rc-button-group"
  >
    <Button
      iconPosition="left"
      size="normal"
      tag="button"
      type="primary"
    >
      <button
        className="rc-button rc-button--primary rc-button--normal"
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="rc-button__content"
        >
          <span
            className="rc-button__text"
            key="text"
          >
            next step
          </span>
        </div>
      </button>
    </Button>
  </div>
</ButtonGroup>
`;

exports[`Button should render correctly when icon is JSX element 1`] = `
<Button
  icon={
    <Icon
      name="shop-o"
      tag="i"
    />
  }
  iconPosition="left"
  size="normal"
  tag="button"
  type="default"
>
  <button
    className="rc-button rc-button--default rc-button--normal"
    onClick={[Function]}
    style={Object {}}
  >
    <div
      className="rc-button__content"
    >
      <Icon
        name="shop-o"
        tag="i"
      >
        <Badge
          className="van-icon van-icon-shop-o"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper van-icon van-icon-shop-o"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
  </button>
</Button>
`;

exports[`Button should render empty button without errors 1`] = `
<button
  class="rc-button rc-button--default rc-button--normal"
>
  <div
    class="rc-button__content"
  >
    <span
      class="rc-button__text"
    />
  </div>
</button>
`;
